### PR TITLE
Port to_padded_tensor CUDA kernel from pytorch/nestedtensor

### PR DESCRIPTION
**Pull request by cpuhrsch**
This PR adds a custom CUDA kernel to pad NestedTensors between dimension 2 and 4.

### CODE REVIEW

File: aten/src/ATen/NestedTensorImpl.h [hunk 1]
```diff
@@ -32,6 +32,9 @@ struct TORCH_API NestedTensorImpl : public c10::TensorImpl {
   const Tensor& get_nested_size_tensor() const {
     return nested_size_tensor_;
   }
+  c10::optional<int64_t> get_opt_size(int64_t i) const {
```
Comment (cpuhrsch) [inline aten/src/ATen/NestedTensorImpl.h:35]: We could make this lazy and only calculate and store once the method was called.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 2]
```diff
@@ -301,7 +277,7 @@ Tensor nested_from_padded_generic(
       std::move(new_buffer), sizes);
 }
 
-Tensor NestedTensor_to_padded_tensor(const Tensor& t, double padding) {
+Tensor NestedTensor_to_padded_tensor_generic(const Tensor& t, double padding) {
   // TODO port CUDA path in pytorch/nestedtensor to_padded_tensor!
```
Comment (swolchok) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:281]: delete TODO

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 3]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+  return result;
+}
+
+Tensor batch_offsets_from_efficient_size(Tensor ef_sizes) {
+  int64_t* nt_sizes_ptr = ef_sizes.data_ptr<int64_t>();
+  Tensor offsets = at::empty({1 + ef_sizes.size(0)}, at::kLong);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+  offsets_ptr[0] = 0;
+  int64_t ef_sizes_size_1 = ef_sizes.size(1);
+  for (int64_t i = 0; i < ef_sizes.size(0); i++) {
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:137]: cache size(0) too, make the locals const, probably have to use c10::irange since we're pytorch core here and somebody will come along and change it if we don't

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 4]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+
+Tensor NestedTensor_to_padded_tensor_cuda(const Tensor& t, double padding) {
+  if ((t.dim() >= 2 && t.dim() <= 4)) {
+    auto orig_nt_dim = t.dim();
+    auto* nt_input = get_nested_tensor_impl(t);
+    TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+    const auto& nt_buffer = nt_input->get_buffer();
+    const auto nt_input_opt_size_2 = nt_input->get_opt_size(2);
+
+    Tensor nt = t;
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:155]: nit: can save a refcount bump using c10::MaybeOwned<Tensor> here

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 5]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+    Tensor output = at::empty(IntArrayRef(new_size), nt_buffer.options());
+
+    int64_t input_dim = nt_sizes.size(1);
+    int64_t batch_size = nt_sizes.size(0);
+    at::Tensor metadata = at::cat({offsets, nt_sizes.reshape(-1)});
+    metadata = metadata.to(at::Device(kCUDA), at::kInt, true, true);
+
+    std::vector<int64_t> split_sizes;
+    split_sizes.push_back(offsets.numel());
+    split_sizes.push_back(nt_sizes.numel());
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:174]: don't heap-allocate unnecessarily; use `std::array<int64_t, 2>` here or just `{offsets.numel(), nt_sizes.numel()}` at the use site instead

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 6]
```diff
@@ -226,5 +226,211 @@ template void remove_padding_transform0213_kernelLauncher<c10::Half>(
...
+    T* output,
+    T padding_value,
+    const int* offsets,
+    const int* input_sizes,
+    int input_dim,
+    int output_sizes_1,
+    const int batch_size)
+{
+  const int batch_id  = blockIdx.x;
+  const int grid_id  = blockIdx.y;
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:243]: nit: extra spaces here

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 7]
```diff
@@ -226,5 +226,211 @@ template void remove_padding_transform0213_kernelLauncher<c10::Half>(
...
+      output[output_offset] = input[batch_input_offset + i];
+    } else {
+      output[output_offset] = padding_value;
+    }
+  }
+}
+
+template<typename T>
+__global__
+void add_padding_2(
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:271]: I suspect there's a clever way to use templates to deduplicate these, but we can always do that later

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 8]
```diff
@@ -226,5 +226,211 @@ template void remove_padding_transform0213_kernelLauncher<c10::Half>(
...
+
+template<typename T>
+void add_padding_kernelLauncher(
+    T* input, // [batch_size x None]
+    T* output, // [batch_size x max(input.nested_size(1)) x inner_size]
+    T padding_value,
+    const int* offsets,
+    const int* input_sizes,
+    int input_dim,
+    std::vector<int64_t> output_sizes,
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:370]: take by reference?

File: aten/src/ATen/NestedTensorImpl.h [hunk 9]
```diff
@@ -63,6 +66,7 @@ struct TORCH_API NestedTensorImpl : public c10::TensorImpl {
 
   at::Tensor buffer_;
   const at::Tensor nested_size_tensor_;
+  std::vector<c10::optional<int64_t>> opt_sizes_;
```
Comment (swolchok) [inline aten/src/ATen/NestedTensorImpl.h:69]: this is not a memory-efficient data structure if any integer value could be used to indicate unoccupied; for example, is -1 a valid value? if not, you could store with -1 internally and convert to c10::optional<int64_t> when returning from get_opt_size

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 10]
```diff
@@ -8,6 +8,32 @@
 namespace at {
 namespace native {
 
+inline std::vector<c10::optional<int64_t>> construct_efficient_size(
+    const at::Tensor& sizes) {
+  std::vector<c10::optional<int64_t>> result;
+  if (sizes.dim() == 0)
+    return result;
```
Comment (swolchok) [inline aten/src/ATen/NestedTensorImpl.cpp:15]: braces please

File: aten/src/ATen/native/nested/NestedTensorTransformerFunctions.h [hunk 11]
```diff
@@ -72,5 +71,16 @@ void remove_padding_transform0213_kernelLauncher(
...
+
+template <typename T>
+void add_padding_kernelLauncher(
+    T* input,
+    T* output,
+    T padding_value,
+    const int* offsets,
+    const int* input_sizes,
+    int input_dim,
+    std::vector<int64_t> output_sizes,
```
Comment (swolchok) [inline aten/src/ATen/native/nested/NestedTensorTransformerFunctions.h:83]: take by reference

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 12]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+    Tensor collapsed_sizes = sizes_dim1 * sizes_dim2;
+    new_nt_sizes = collapsed_sizes.contiguous();
+  }
+  Tensor result = at::detail::make_tensor<NestedTensorImpl>(nt_input->get_buffer(), new_nt_sizes);
+  TORCH_CHECK(result.dim() == 2, "Expected result to be 2 dimensional.");
+  return result;
+}
+
+Tensor batch_offsets_from_efficient_size(Tensor ef_sizes) {
+  int64_t* nt_sizes_ptr = ef_sizes.data_ptr<int64_t>();
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:132]: would prefer sizes()[x] over size(x) whenever dim is known because it's more efficient

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 13]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+    nt_input = get_nested_tensor_impl(nt);
+    Tensor nt_sizes = nt_input->get_nested_size_tensor();
+    Tensor offsets = batch_offsets_from_efficient_size(nt_sizes);
+    auto new_size = NestedTensor_get_max_size(*nt_input);
+    new_size.insert(new_size.begin(), nt_sizes.size(0));
+    Tensor output = at::empty(IntArrayRef(new_size), nt_buffer.options());
+
+    int64_t input_dim = nt_sizes.size(1);
+    int64_t batch_size = nt_sizes.size(0);
+    at::Tensor metadata = at::cat({offsets, nt_sizes.reshape(-1)});
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:169]: we do some clever stuff in other paths to avoid the at::cat here; can we at least have a TODO to replicate it here?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 14]
```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
...
+  TORCH_CHECK(dim1 > 0, "dim1: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 > 0, "dim2: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 - 1 == dim1, "dim2 must be one more than dim1.")
+  TORCH_CHECK(dim1 == 1, "dim1 must be 1.")
+  TORCH_CHECK(input.dim() == 3, "Expected input to be 3 dim.");
+
+  auto* nt_input = get_nested_tensor_impl(input);
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+  const auto& input_buffer = nt_input->get_buffer();
```
Comment (mjanderson09) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:116]: unused variable "input_buffer"

File: aten/src/ATen/NestedTensorImpl.h (hunk 1 above)
Comment (swolchok) [reply aten/src/ATen/NestedTensorImpl.h:35]: I assume the description is "analogous to `Tensor::size()`, but only applies if all sub-tensors (or however you want to word it) have the same size in the ith dimension". Would help to write that. Also, I think it should be called `opt_size` to match `Tensor::size()`?

File: aten/src/ATen/native/nested/NestedTensorMath.h [hunk 15]
```diff
@@ -7,7 +7,32 @@ struct NestedTensorImpl;
...
-std::vector<int64_t> NestedTensor_get_max_size(const NestedTensorImpl& nt);
+inline std::vector<int64_t> NestedTensor_get_max_size_from_size_tensor(const Tensor& sizes) {
+  if (sizes.dim() == 0) {
+    return {};
+  }
+  const auto sizes_ptr = sizes.data_ptr<int64_t>();
+  const auto sizes_size_0 = sizes.sizes()[0];
+  const auto sizes_size_1 = sizes.sizes()[1];
+  TORCH_INTERNAL_ASSERT(sizes_size_1 > 0);
+  std::vector<int64_t> results(sizes_size_1, 0);
```
Comment (swolchok) [inline aten/src/ATen/native/nested/NestedTensorMath.h:18]: shouldn't this be a DimVector? IIUC it's got expected length the same as sizes() of some Tensor. (we can make this a separate PR; I realize this is probably a move of stuff I did)

Comment (swolchok) [reply aten/src/ATen/native/nested/NestedTensorMath.h:18]: Also, this seems like a pretty large function. Why is it inline in a header?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 16]
```diff
@@ -101,5 +102,106 @@ Tensor nested_from_padded_cuda(
...
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+
+  Tensor sizes_dim1 = at::native::narrow(nt_sizes, 1, 0, 1);
+  Tensor sizes_dim2 = at::native::narrow(nt_sizes, 1, 1, 1);
+
+  Tensor new_nt_sizes;
+  if (dim1 == 1) {
+    Tensor collapsed_sizes = sizes_dim1 * sizes_dim2;
+    new_nt_sizes = collapsed_sizes.contiguous();
```
Comment (swolchok) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:123]: why wouldn't it be contiguous already? it's a newly-allocated Tensor

File: test/test_nestedtensor.py [hunk 17]
```diff
@@ -272,14 +273,50 @@ def test_to_padded_tensor_simple(self, device):
...
+            torch.randn(2, 3, device=device, dtype=dtype),
+            torch.randn(4, 5, device=device, dtype=dtype),
         ]
-        nt = torch.nested_tensor(ts, device=device)
+        nt = torch.nested_tensor(ts, device=device, dtype=dtype)
+        pad = 42
+        correct_output = []
+        for i in range(len(ts)):
+            correct_output.append(torch.ones_like(ts[2]) * pad)
+            correct_output[i][:ts[i].size(0), :ts[i].size(1)].copy_(ts[i])
```
Comment (swolchok) [inline test/test_nestedtensor.py:307]: you don't need an index here

```
for t in ts:
    next = torch.ones_like(ts[2]) * pad
    correct_output.append(next)
    next[:t.size(0), :t.size(1)].copy_(t)

File: aten/src/ATen/native/nested/NestedTensorMath.h (hunk 15 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/NestedTensorMath.h:18]: I couldn't get it to link correctly if defined in NestedTensorMath.cpp, so I gave up and moved it into the header. Not an ideal solution.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 16 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:123]: Good catch, this is a relic from pytorch/nestedtensor

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 18]
```diff
@@ -101,5 +102,102 @@ Tensor nested_from_padded_cuda(
     return at::native::nested_from_padded_generic(padded, sizes);
   }
 }
+
+Tensor _collapse_two_dims_3(const Tensor& input, int64_t dim1, int64_t dim2) {
+  TORCH_CHECK(dim1 > 0, "dim1: Cannot collapse dim 0.");
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:107]: do you expect to accept different args in future? Now it looks like the only valid values are `1` and `2` for dim1 and dim2, so doesn't make sense to pass them as args.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 19]
```diff
@@ -101,5 +102,102 @@ Tensor nested_from_padded_cuda(
...
+  TORCH_CHECK(dim2 - 1 == dim1, "dim2 must be one more than dim1.")
+  TORCH_CHECK(dim1 == 1, "dim1 must be 1.")
+  TORCH_CHECK(input.dim() == 3, "Expected input to be 3 dim.");
+
+  auto* nt_input = get_nested_tensor_impl(input);
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+
+  Tensor sizes_dim1 = at::native::narrow(nt_sizes, 1, 0, 1);
+  Tensor sizes_dim2 = at::native::narrow(nt_sizes, 1, 1, 1);
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:118]: when you are calling this function, all `sizes_dim2` are the same, so you don't need to do `narrow` here and mutliply tensor by tensor? Multiplying `sizes_dim1` by constant will do.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 20]
```diff
@@ -101,5 +102,102 @@ Tensor nested_from_padded_cuda(
...
+    Tensor offsets = batch_offsets_from_efficient_size(nt_sizes);
+    auto new_size = NestedTensor_get_max_size(*nt_input);
+    new_size.insert(new_size.begin(), nt_sizes.sizes()[0]);
+    Tensor output = at::empty(IntArrayRef(new_size), nt_buffer.options());
+
+    int64_t input_dim = nt_sizes.sizes()[1];
+    int64_t batch_size = nt_sizes.sizes()[0];
+    // TODO: Remove need for cat here
+    at::Tensor metadata = at::cat({offsets, nt_sizes.reshape(-1)});
+    metadata = metadata.to(at::Device(kCUDA), at::kInt, true, true);
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:167]: Doing async transfer here is dangerous because metadata may go out of scope before actual transfer happens

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 21]
```diff
@@ -101,5 +102,102 @@ Tensor nested_from_padded_cuda(
...
+      prod = prod * nt_sizes_ptr[i * ef_sizes_size_1 + j];
+    }
+    offsets_ptr[i + 1] = offsets_ptr[i] + prod;
+  }
+  return offsets;
+}
+
+Tensor NestedTensor_to_padded_tensor_cuda(const Tensor& t, double padding) {
+  int64_t t_dim = t.dim();
+  if ((t_dim >= 2 && t_dim <= 4)) {
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:145]: should you also check dtypes here?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 22]
```diff
@@ -226,5 +226,207 @@ template void remove_padding_transform0213_kernelLauncher<c10::Half>(
...
+    const int* offsets,
+    const int* input_sizes,
+    int input_dim,
+    int output_sizes_1,
+    int output_sizes_2,
+    const int batch_size) {
+  const int batch_id = blockIdx.x;
+  const int grid_id = blockIdx.y;
+  const int tid = threadIdx.x + grid_id * 256;
+  const int grainsize = 16 * 256;
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:281]: YOu might want to constexpr those and use them in grid dimension also, right now blockSize and gridSize have to match these hardcoded values (same for all other kernels)

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 23]
```diff
@@ -226,5 +226,207 @@ template void remove_padding_transform0213_kernelLauncher<c10::Half>(
...
+  const int offset = offsets[batch_id];
+  const int* sizes_i = input_sizes + batch_id * input_dim;
+  const int output_offset = batch_id * output_sizes_1 * output_sizes_2;
+  const int output_numel = output_sizes_1 * output_sizes_2;
+  for (int ii = 0; ii < (output_numel / grainsize); ii++) {
+    const int i = ii * grainsize + tid;
+    const int i0 = i / (output_sizes_2);
+    const int i1 = i % output_sizes_2;
```
Comment (ngimel) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:289]: i - i0 * output_sizes_2

File: test/test_nestedtensor.py [hunk 24]
```diff
@@ -272,14 +273,52 @@ def test_to_padded_tensor_simple(self, device):
...
+        self.assertEqual(padded, correct_output)
 
-    def test_to_padded_tensor_unrelated_shapes(self, device):
+    @dtypes(torch.float, torch.float16)
+    def test_to_padded_tensor_dim3(self, device, dtype):
         ts = [
-            torch.randn(1, 2, 3, device=device),
-            torch.randn(2, 3, 4, device=device),
-            torch.randn(4, 5, 6, device=device),
+            torch.randn(1, 2, device=device, dtype=dtype),
```
Comment (ngimel) [inline test/test_nestedtensor.py:299]: you might want to test bigger inputs, so far the "vectorized" part of your loops are never tested.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 18 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:107]: Yes, I do. It's probably better to make these comments...

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 21 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:145]: Yes, it'd be a more efficient fallback.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu (hunk 23 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:289]: Is this a generic optimization that works for modulus operations? I assume this is much more efficient from what I've read up on modulus.

File: aten/src/ATen/native/nested/NestedTensorMath.h (hunk 15 above)
Comment (swolchok) [reply aten/src/ATen/native/nested/NestedTensorMath.h:18]: TORCH_API didn't help?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu (hunk 22 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:281]: Since this applies to the whole file I'd rather do this in a follow up.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 25]
```diff
@@ -107,5 +108,78 @@ Tensor nested_from_padded_cuda(
...
+
+Tensor NestedTensor_to_padded_tensor_cuda(const Tensor& t, double padding) {
+  int64_t t_dim = t.dim();
+  if (t_dim >= 2 && t_dim <= 4 &&
+      (t.dtype() == at::kFloat || t.dtype() == at::kDouble ||
+       t.dtype() == at::kHalf)) {
+    auto* nt_input = get_nested_tensor_impl(t);
+    TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+    const auto& nt_buffer = nt_input->get_buffer();
+    const auto nt_input_opt_size_2 = nt_input->opt_size(2);
```
Comment (mjanderson09) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:137]: Can this possibly exceed the dimension of opt_sizes_ in the case that dim is 2? For example in the second recursive call to this fn after the narrowing clause below.

**Review by ngimel — APPROVED** [review]
Cool, please do a follow up to make sure grid/block dims in the launcher and in the kernel match.
